inbox: Add "Inbox" page!

chrisbobbe opened the pull request:
Stacked atop #371.

Still a draft for now, but wanted to share what I have so far. 🙂

chrisbobbe commented:
#380 in its current form would fix a small deviation from the Figma. No need to have it block progress on this though.

gnprice commented:
(Cross-linking: this fixes #117, right?)

gnprice commented on lib/widgets/inbox.dart:
```diff
@@ -0,0 +1,518 @@
+  const StreamSectionData(this.streamId, this.count, this.items);
+}
+
+class AllDmsHeaderItem extends StatelessWidget {
```
Can these widget classes be private?

If they remain public, they should get more specific names that make it clear they're for the inbox. E.g., `InboxAllDmsHeaderItem`. That will mitigate having them show up in IDE autocomplete when working in unrelated parts of the code.

gnprice reviewed (commented):
Thanks! Generally looks good; comments below.

A couple of comments are about efficiency or getting work out of the build methods. But I want to highlight this high-order bit about that aspect:
> As long as this feels reasonably responsive in manual testing, though, then the right thing for now is to prioritize speed of writing the code over speed of running it, and just leave a TODO.

Thoughts on the TODO items in the main commit's message:

    TODO (before merge?):
    
    - tests

Yeah. For now let's keep the tests simple, just in the interest of prioritizing our other Alpha and Beta 1 work. But it'd be good to have (a) a simple smoke test where there's some unread DMs and some unread stream conversations; and (b) a quick test of the collapse/uncollapse functionality.

(Fancier tests would include that if you collapse one section, then scroll it far offscreen and scroll it back on, it's still collapsed. That's something that's significant to the UX but would fail in a plausible alternate implementation, where the collapsed state was local to the list item for the section.)

    - sorting (of streams; and within a stream, of topics)

These are OK to leave as TODOs if adding them to the implementation seems complicated.

    - icons:
        https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/design.3A.20.23F117.20.22Inbox.22.20screen/near/1680637
      (that message, and the next)

Will be good to fix but doesn't need to block merging; if we merge before fixing those, we can file follow-up issues for Beta 2.

    - currently the whole stream row is tappable to collapse/uncollapse;
      should we limit that to just the area near the
      arrow_down/arrow_right, and have the rest of the row open a
      message list to the stream narrow?

Yeah, I'd be inclined to do the latter. If the Figma design expresses the former (or if this needs further discussion for another reason), then this would be a good question for that chat thread.


    - @-mention indicator next to stream/topic unread badge count

Let's file a Beta 2 followup issue for this. Can sort it toward the top of Beta 2 on the priorities board.

    - helper widgets; code structure etc.?

Comments on this below.

    - commit msg. with "fixes" line

gnprice commented on test/example_data.dart:
```diff
@@ -123,6 +123,51 @@ ZulipStream stream({
+    isMuted: isMuted ?? true,
+    isWebPublic: isWebPublic ?? true,
+    color: color ?? 0xffe79ab5,
+    streamPostPolicy: streamPostPolicy ?? 1,
+    messageRetentionDays: messageRetentionDays ?? 365,
+    historyPublicToSubscribers: historyPublicToSubscribers ?? true,
+    firstMessageId: firstMessageId ?? 123, // TODO generate example data
+    streamWeeklyTraffic: streamWeeklyTraffic ?? 100, // TODO generate example data
+    canRemoveSubscribersGroupId: canRemoveSubscribersGroupId ?? 123 // TODO generate example data
```
(This may end up coming first from #382.)

chrisbobbe commented on lib/widgets/inbox.dart:
```diff
@@ -0,0 +1,518 @@
+  State<InboxPage> createState() => InboxPageState();
+}
+
+class InboxPageState extends State<InboxPage> with PerAccountStoreAwareStateMixin<InboxPage> {
```
Sure, it can be `@visibleForTesting`. If I make it private, by prefixing with `_`, I get analysis feedback saying "Invalid use of a private type in a public API" in several places. Such as:

```dart
class AllDmsHeaderItem extends StatelessWidget {
  const AllDmsHeaderItem({
    super.key,
    required this.count,
    required this.collapsed,
    required this.pageState, // <-- here
  });

  final int count;
  final bool collapsed;
  final _InboxPageState pageState; // <-- here
```

gnprice replied:
Ah, I guess that points toward this other comment then :slightly_smiling_face: : https://github.com/zulip/zulip-flutter/pull/381#discussion_r1391752223

chrisbobbe commented on lib/widgets/inbox.dart:
```diff
@@ -0,0 +1,518 @@
+  @override
+  void onNewStore() {
+    final newStore = PerAccountStoreWidget.of(context);
+    unreadsModel?.removeListener(_modelChanged);
+    unreadsModel = newStore.unreads..addListener(_modelChanged);
+    recentDmConversationsModel = newStore.recentDmConversationsView
+      ..addListener(_modelChanged);
```
Yep, thanks for the catch!

chrisbobbe commented on lib/widgets/inbox.dart:
```diff
@@ -0,0 +1,518 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return Material(
+      color: collapsed ? Colors.white : const Color(0xFFF3F0E7),
+      child: InkWell(
+        onTap: () {
+          pageState.allDmsCollapsed = !collapsed;
+        },
+        // TODO min-height 48px, like other touch targets?
+        child: Row(crossAxisAlignment: CrossAxisAlignment.center, children: [
+          Padding(padding: const EdgeInsets.all(10),
+            child: Icon(size: 20, color: const Color(0x7F1D2E48),
+              collapsed ? ZulipIcons.arrow_right : ZulipIcons.arrow_down)),
+          const Icon(size: 18, color: Color(0xFF222222),
+            ZulipIcons.user),
+          const SizedBox(width: 5),
+          Expanded(child: Padding(
+            padding: const EdgeInsets.symmetric(vertical: 4),
+            child: Text(
+              style: const TextStyle(
+                fontFamily: 'Source Sans 3',
+                fontSize: 17,
+                height: (20 / 17),
+                color: Color(0xFF222222),
+              ).merge(weightVariableTextStyle(context, wght: 600, wghtIfPlatformRequestsBold: 900)),
+              maxLines: 1,
+              overflow: TextOverflow.ellipsis,
+              'Direct messages'))),
+          const SizedBox(width: 12),
+          Padding(padding: const EdgeInsetsDirectional.only(end: 16),
+            child: UnreadCountBadge(baseStreamColor: null, bold: true,
+              count: count)),
```
Would the shared helper `_HeaderItemLayout` take an `int` param for the unread count?

gnprice replied:
Ah yeah, that too. And I guess the `onTap` callback as well.

chrisbobbe replied:
I wasn't seeing a smooth path to the first way, so in my next revision I do something I found easy enough. It's basically option 3 (like the abstract `MessageActionSheetMenuItemButton` and its subclasses in lib/widgets/action_sheet.dart), except the awkward bit is solved. The icon color and bar color do share computation of the clamped color. As a bonus, all the unread-count badges share computation of their color, too. (There are a lot of those, when a stream has unreads across many topics.) A swatch of colors related to the stream color is memoized on the `Subscription` object.

chrisbobbe commented:
Thanks for the review! Revision pushed, and please see my reply at https://github.com/zulip/zulip-flutter/pull/381#discussion_r1396344813.

gnprice reviewed (commented):
Thanks! A few comments below, but otherwise this all looks good — it just needs tests. (Which can be kept pretty simple, as described above at https://github.com/zulip/zulip-flutter/pull/381#pullrequestreview-1728251814 .)

We've talked about several TODO items here that you're going to file as issues for Beta 2.

---

As I mentioned in our call today, I played with this on my Pixel 5 to see the performance, using a test user on chat.zulip.org who has a ton of unreads. In debug mode the performance was pretty choppy — occasional pauses while scrolling that I believe reflected a new section scrolling into view, so that with the current version the whole section has to be built and laid out, which can be hundreds of items for different topics. Those pauses in debug mode were sometimes around a second long.

It's fine for debug mode to be choppy, though, so I tried it in profile mode. There there's still enough jank to be noticeable, but it's a lot milder — mostly a couple of dropped frames at a time, sometimes like 150ms. So this level of performance is fine for Beta 1, but it's something we'll want to revisit for Beta 2.

Probably that means I'll rework something in StickyHeaderListView to make it possible to have a separate list item for each conversation, just like you did in a previous draft, and without the glitchy interactions that that led to between one section's header and the next. (Effectively the current StickyHeaderListView assumes that each header is no taller than the item it belongs to; that's true in the message list, and it's true here by making each entire section into an item, but it's not true if each conversation is its own item.)

chrisbobbe commented on lib/widgets/unread_count_badge.dart:
```diff
@@ -13,57 +13,34 @@ class UnreadCountBadge extends StatelessWidget {
 
   @override
   Widget build(BuildContext context) {
+    final effectiveBackgroundColor = switch (backgroundColor) {
+      // TODO why is the "as" casting needed?
+      StreamColorSwatch() => (backgroundColor as StreamColorSwatch).unreadCountBadgeBackground,
```
Makes sense! Thanks.

chrisbobbe commented:
Thanks for the review! Revision pushed. Oh and here's a screenshot; @alya and @terpimost FYI:

<img width="498" alt="Screenshot 2023-11-17 at 10 51 30 PM" src="https://github.com/zulip/zulip-flutter/assets/22248748/08939272-96e0-4506-a338-87875799e926">

-----

And for Greg:

- I rebased past #382 (which helpfully made `PerAccountStoreTestExtension.addSubscriptions` easy to implement; thanks @sirpengi!)
- I did a first pass at ordering streams and topics, and I don't _think_ it caused significant extra lag in my manual testing, but definitely will be good to get that work out of the build method.
- I added and used one more color on `StreamColorSwatch`: the one named `iconOnPlainBackground`. (I think I renamed some others since the last revision.)
- I added some simple tests. I tried to write them so that they wouldn't break when we reimplement with separate `StickyHeaderListView` items per topic and DM conversation (#389). In particular, I didn't give them access to our custom widgets like `_StreamHeaderItem` and `_TopicItem`. It might help to give the tests that access eventually; e.g., the recent-DMs widget tests are helped in their work by having access to `RecentDmConversationsItem`. But I felt like it might be good to hold off for now, especially with #389 on the horizon.

gnprice commented on lib/widgets/inbox.dart:
```diff
@@ -0,0 +1,513 @@
+      topicItems.sort((a, b) {
+        final (_, _, aLastUnreadId) = a;
+        final (_, _, bLastUnreadId) = b;
+        return bLastUnreadId - aLastUnreadId;
```
… I guess here's one good reason to think overflow will never happen: these message IDs are all over the web client, represented there as plain JS numbers. So they need to fit into 54 bits in order to not break web.

Still, `compareTo` feels like it says a bit more directly what we mean here.

gnprice reviewed (commented):
Thanks! This generally all looks good; a couple of comments below.

Since this is so close, in the interest of efficiency (and since you're focusing on the next feature, the reactions UI), I'll just make the tweaks to fix these and merge.

chrisbobbe commented on test/api/model/model_test.dart:
```diff
@@ -102,6 +103,222 @@ void main() {
+        // in <https://replit.com/@VladKorobov/zulip-topic-feed-colors#script.js>.
+        // (Skipping `streamColors` because there are 100+ of them.)
+        // On how to extract expected results from the replit, see:
+        //   TODO
```
Oh! This time I did save my work. 🙂 I meant to come back to this before submitting my revision; oops.

Here's the code I added in the [replit](https://replit.com/@VladKorobov/zulip-topic-feed-colors#script.js), just under `ZULIP_ASSIGNMENT_COLORS`. It duplicates some of the logic, where it wasn't exposed with a convenient helper function. But it can be checked against the implementation.

<details>

```js
const dartColorIntFromHexString = (hexString) => {
  const lowercaseWithoutHash = hexString.substring(1).toLowerCase();
  switch (lowercaseWithoutHash.length) {
    case 6: // opaque (alpha omitted)
      return `0xff${lowercaseWithoutHash}`;
    case 8: // not opaque
      // convert rrggbbaa to aarrggbb
      return '0x'
        + lowercaseWithoutHash.substring(6, 8)
        + lowercaseWithoutHash.substring(0, 2);
    default:
      throw new Error();
  }
};

const dartCheckLine = (baseHex, expectedHex) => {
  return `runCheck(${dartColorIntFromHexString(baseHex)}, const Color(${dartColorIntFromHexString(expectedHex)}));\n`;
};

let iconOnPlainBackgroundResult = '';
let iconOnBarBackgroundResult = '';
let barBackgroundResult = '';
ZULIP_ASSIGNMENT_COLORS.forEach(color => {
  const correctedColor = correctStreamColor({ color });
  const barBackground = getRecipientBarColor({ color: correctStreamColor({ color })});
  const iconOnPlainBackground = correctedColor;
  const iconOnBarBackground = colord(correctedColor).darken(0.12).toHex();

  iconOnPlainBackgroundResult += dartCheckLine(color, iconOnPlainBackground);
  iconOnBarBackgroundResult += dartCheckLine(color, iconOnBarBackground);
  barBackgroundResult += dartCheckLine(color, barBackground);
});

console.log(iconOnPlainBackgroundResult);
console.log(iconOnBarBackgroundResult);
console.log(barBackgroundResult);
```
</details>

chrisbobbe replied:
Pushed 640caef41e5f52076e49a24eb119bbbe0ae2691c linking to my message here with that code.

gnprice replied:
Great, thanks!

chrisbobbe commented on lib/widgets/inbox.dart:
```diff
@@ -0,0 +1,513 @@
+        // TODO(i18n) something like JS's String.prototype.localeCompare
+        return streamA.name.toLowerCase().compareTo(streamB.name.toLowerCase());
+      })
+      ..sort((a, b) {
```
Ah OK, makes sense.

chrisbobbe commented:
Thanks for the review and those fixes!
